Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp

```diff
@@ -123,9 +295,35 @@ void URobotVehicleMovementComponent::UpdateOdom(float InDeltaTime)
     OdomData.pose_pose_position_z = Pos.Z;
     OdomData.pose_pose_orientation = Rot;
 
-    OdomData.twist_twist_linear = OdomData.pose_pose_orientation.UnrotateVector(Pos - PreviousEstimatedPos) / InDeltaTime;
-    OdomData.twist_twist_angular =
-        FMath::DegreesToRadians((Rot * PreviousEstimatedRot.Inverse()).GetNormalized().Euler()) / InDeltaTime;
+    // compute velocities values with real position and orientation of robot
+    FQuat CurrentOrientation = FQuat(PawnOwner->GetActorRotation() - InitialTransform.GetRotation().Rotator());
+	FQuat DeltaQ = CurrentOrientation * LastOrientation.Inverse();
+    FVector NewAngularVelocity = FMath::DegreesToRadians( DeltaQ.Euler() ) / InDeltaTime;
+
+    // linear velocity is expressed in robot local coordinates, taking into account the robot rotation
+    FVector NewVelocity = OdomData.pose_pose_orientation.UnrotateVector( Pos-LastLocation ) / InDeltaTime;
+
+	LastOrientation = CurrentOrientation;
+    LastLocation = Pos;
+
+    // velocity
+    OdomData.twist_twist_linear = NewVelocity;
+    OdomData.twist_twist_angular = NewAngularVelocity;
+}
+
+void URobotVehicleMovementComponent::UpdateEntityState( FString AgentName )
+{
+    EntityState.name = AgentName + TEXT("/base_footprint");
+    EntityState.pose_position_x = OdomData.pose_pose_position_x / 100.f;
+    EntityState.pose_position_y = OdomData.pose_pose_position_y / 100.f;
+    EntityState.pose_position_z = OdomData.pose_pose_position_z / 100.f;
+    EntityState.pose_orientation = OdomData.pose_pose_orientation;
+    EntityState.pose_position_y = -EntityState.pose_position_y;
+    EntityState.pose_orientation.X = -EntityState.pose_orientation.X;
+    EntityState.pose_orientation.Z = -EntityState.pose_orientation.Z;

```

Comment:
`const FString& InAgentName`
Should we use the conversion utils from UE to ROS here?